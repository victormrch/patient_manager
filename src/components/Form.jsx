const Form = () => {
  return (
    <div className="md:w-1/2 lg:w-2/5">
      <h2 className="font-black text-3xl text-center">Patient Follow-Up</h2>
      <p className="text-lg mt-5 text-center mb-6">
        Add Patient and {""}
        <span className="text-indigo-600 font-bold ">Manage Them</span>
      </p>
      <form className="bg-white shadow-md rounded-lg py-10 px-5 mb-10">
        <div className="mb-5">
          <label
            htmlFor="pet"
            className="block text-gray-700 uppercase font-bold"
          >
            Pet Name
          </label>
          <input
            id="pet"
            type="text"
            placeholder="Enter the Pet's Name"
            className="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          />
        </div>
        <div className="mb-5">
          <label
            htmlFor="owner"
            className="block text-gray-700 uppercase font-bold"
          >
            Owner Name
          </label>
          <input
            id="owner"
            type="text"
            placeholder="Enter the Owner's Name"
            className="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          />
        </div>
        <div className="mb-5">
          <label
            htmlFor="email"
            className="block text-gray-700 uppercase font-bold"
          >
            Email Address
          </label>
          <input
            id="email"
            type="email"
            placeholder="Enter the owner's Mail"
            className="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          />
        </div>
        <div className="mb-5">
          <label
            htmlFor="date"
            className="block text-gray-700 uppercase font-bold"
          >
            discharge date
          </label>
          <input
            id="date"
            type="date"
            className="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          />
        </div>
        <div className="mb-5">
          <label
            htmlFor="symptoms"
            className="block text-gray-700 uppercase font-bold"
          >
            Reason for visit
          </label>
          <textarea
            id="symptoms"
            className="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
            placeholder="Enter the Pet Symptoms"
          ></textarea>
        </div>
        <input
          type="submit"
          className="bg-indigo-600 w-full p-3 text-white uppercase font-bold hover:bg-indigo-700 cursor-pointer transition-colors"
          value="Add Patient"
        />
      </form>
    </div>
  );
};

export default Form;
