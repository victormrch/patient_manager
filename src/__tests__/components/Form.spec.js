import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import Form from "../../components/Form";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";

const mockEmptyPatient = {
  dischargeDate: "",
  email: "",
  id: "",
  name: "",
  owner: "",
  reason: "",
};

const mockPatient = {
  dischargeDate: "2019-01-01",
  email: "test@test.com",
  id: "1234",
  name: "Test Name",
  owner: "Test Owner",
  reason: "Test Reason",
};

describe("Form component specs", () => {
  it("should render correctly using snapshot testing and botton text should put Add Patient when patient is empty", () => {
    const { asFragment } = render(<Form patient={mockEmptyPatient} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render correctly using snapshot testing and botton text should put Edit Patient when patient is filled", () => {
    const { asFragment } = render(<Form patient={mockPatient} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should display Form component", () => {
    render(<Form patient={mockEmptyPatient} />);

    const elementTitle = screen.getByRole("heading", {
      name: "Patient Follow-Up",
    });
    const elementSubtitle = screen.getByText("Add Patient and");

    expect(elementTitle).toBeInTheDocument();
    expect(elementTitle.tagName).toBe("H2");
    expect(elementTitle.textContent).toBe("Patient Follow-Up");

    expect(elementSubtitle).toBeInTheDocument();
    expect(elementSubtitle.tagName).toBe("P");
    expect(elementSubtitle.textContent).toBe("Add Patient and Manage Them");
  });
  it("when Add Patient button is clicked handleSubmit should be called one time and appear an error is form is empty", () => {
    render(<Form patient={mockEmptyPatient} />);

    const elementAddPatientButton = screen.getByDisplayValue("Add Patient");

    userEvent.click(elementAddPatientButton);

    const elementError = screen.getByText("All fields are required");

    expect(elementError).not.toBeNull();
    expect(elementError).toBeInTheDocument();
    expect(elementError.tagName).toEqual("P");
    expect(elementError.textContent).toBe("All fields are required");
  });
  it("when Add Patient button is clicked handleSubmit should be called one time and send Patient when form is filled", () => {
    const props = {
      patient: mockEmptyPatient,
      setPatient: jest.fn(),
      patients: [],
      setPatients: jest.fn(),
    };

    render(<Form {...props} />);

    userEvent.type(screen.getByLabelText("Pet Name"), "Test Pet Name");
    userEvent.type(screen.getByLabelText("Owner Name"), "Test Owner Name");
    userEvent.type(screen.getByLabelText("Email Address"), "test@email.com");
    userEvent.type(screen.getByLabelText("discharge date"), "2019-01-01");
    userEvent.type(
      screen.getByLabelText("Reason for visit"),
      "Test Reason for visit"
    );

    const elementAddPatientButton = screen.getByDisplayValue("Add Patient");

    userEvent.click(elementAddPatientButton);

    const elementError = screen.queryByText("All fields are required");

    expect(elementError).toBeNull();
    expect(elementError).not.toBeInTheDocument();
  });

  it("when Edit Patient button is clicked handleSubmit should be called one time and send Patient when form is filled", () => {
    const props = {
      patient: mockPatient,
      setPatient: jest.fn(),
      patients: [],
      setPatients: jest.fn(),
    };

    render(<Form {...props} />);

    userEvent.type(screen.getByLabelText("Pet Name"), "Test Pet Name");
    userEvent.type(screen.getByLabelText("Owner Name"), "Test Owner Name");
    userEvent.type(screen.getByLabelText("Email Address"), "test@email.com");
    userEvent.type(screen.getByLabelText("discharge date"), "2019-01-01");
    userEvent.type(
      screen.getByLabelText("Reason for visit"),
      "Test Reason for visit"
    );

    const elementAddPatientButton = screen.getByDisplayValue("Edit Patient");

    userEvent.click(elementAddPatientButton);

    const elementError = screen.queryByText("All fields are required");

    expect(elementError).toBeNull();
    expect(elementError).not.toBeInTheDocument();
  });
});
