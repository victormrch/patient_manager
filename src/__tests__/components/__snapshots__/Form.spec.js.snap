// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component specs should render correctly using snapshot testing and botton text should put Add Patient when patient is empty 1`] = `
<DocumentFragment>
  <div
    class="md:w-1/2 lg:w-2/5 mx-5"
  >
    <h2
      class="font-black text-3xl text-center"
    >
      Patient Follow-Up
    </h2>
    <p
      class="text-lg mt-5 text-center mb-10"
    >
      Add Patient and 
      <span
        class="text-indigo-600 font-bold "
      >
        Manage Them
      </span>
    </p>
    <form
      class="bg-white shadow-md rounded-lg py-10 px-5 mb-10"
    >
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="pet"
        >
          Pet Name
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="pet"
          placeholder="Enter the Pet's Name"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="owner"
        >
          Owner Name
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="owner"
          placeholder="Enter the Owner's Name"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="email"
        >
          Email Address
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="email"
          placeholder="Enter the owner's Mail"
          type="email"
          value=""
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="date"
        >
          discharge date
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="date"
          type="date"
          value=""
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="symptoms"
        >
          Reason for visit
        </label>
        <textarea
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="symptoms"
          placeholder="Enter the Pet Symptoms"
        />
      </div>
      <input
        class="bg-indigo-600 w-full p-3 text-white uppercase font-bold hover:bg-indigo-700 cursor-pointer transition-colors rounded-lg"
        type="submit"
        value="Add Patient"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`Form component specs should render correctly using snapshot testing and botton text should put Edit Patient when patient is filled 1`] = `
<DocumentFragment>
  <div
    class="md:w-1/2 lg:w-2/5 mx-5"
  >
    <h2
      class="font-black text-3xl text-center"
    >
      Patient Follow-Up
    </h2>
    <p
      class="text-lg mt-5 text-center mb-10"
    >
      Add Patient and 
      <span
        class="text-indigo-600 font-bold "
      >
        Manage Them
      </span>
    </p>
    <form
      class="bg-white shadow-md rounded-lg py-10 px-5 mb-10"
    >
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="pet"
        >
          Pet Name
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="pet"
          placeholder="Enter the Pet's Name"
          type="text"
          value="Test Name"
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="owner"
        >
          Owner Name
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="owner"
          placeholder="Enter the Owner's Name"
          type="text"
          value="Test Owner"
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="email"
        >
          Email Address
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="email"
          placeholder="Enter the owner's Mail"
          type="email"
          value="test@test.com"
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="date"
        >
          discharge date
        </label>
        <input
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="date"
          type="date"
          value="2019-01-01"
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          class="block text-gray-700 uppercase font-bold"
          for="symptoms"
        >
          Reason for visit
        </label>
        <textarea
          class="border-2 w-full p-2 mt-w placeholder-gray-400 rounded-md"
          id="symptoms"
          placeholder="Enter the Pet Symptoms"
        >
          Test Reason
        </textarea>
      </div>
      <input
        class="bg-indigo-600 w-full p-3 text-white uppercase font-bold hover:bg-indigo-700 cursor-pointer transition-colors rounded-lg"
        type="submit"
        value="Edit Patient"
      />
    </form>
  </div>
</DocumentFragment>
`;
